File: bot-6.js
class TradingBot {
  initialize(state, context, args) {
    state.position = null;
    state.entryPrice = null;
    state.apiCounter = 15; // Initialize with maximum allowed API counter
    state.lastApiCall = Date.now(); // Track the last API call time
  }

  async runIteration(state, context, args) {
    const waitForApiLimit = async () => {
      let elapsedTime = (Date.now() - state.lastApiCall) / 1000;
      state.apiCounter += elapsedTime * 0.33; // Increment counter based on elapsed time
      state.apiCounter = Math.min(state.apiCounter, 15); // Cap the counter at 15
      while (state.apiCounter < 1) {
        // Ensure at least one API call is available
        await new Promise((resolve) => setTimeout(resolve, 1000));
        elapsedTime = (Date.now() - state.lastApiCall) / 1000;
        state.apiCounter += elapsedTime * 0.33;
        state.apiCounter = Math.min(state.apiCounter, 15);
      }
      state.apiCounter -= 1; // Decrement the counter for the current API call
      state.lastApiCall = Date.now();
    };

    try {
      const symbol = args.params.pair;
      const capitalBase = parseFloat(args.params.capitalBase);
      await waitForApiLimit();
      const balances = await this.exchange.fetchBalance();
      const usdtBalance = balances.USDT.free;

      if (state.position === null && usdtBalance >= capitalBase) {
        // Buy ETH/USDT using 100% of starting capital
        await waitForApiLimit();
        const ticker = await this.exchange.fetchTicker(symbol);
        const ethAmount = capitalBase / ticker.last;
        await waitForApiLimit();
        const order = await this.exchange.createOrder(symbol, "market", "buy", ethAmount);
        state.position = order.id;
        state.entryPrice = ticker.last;
        console.log(`Bought ${ethAmount} ETH at ${state.entryPrice} USDT`);
      } else if (state.position !== null) {
        // Check if we can sell at 2% profit
        await waitForApiLimit();
        const ticker = await this.exchange.fetchTicker(symbol);
        const targetPrice = state.entryPrice * 1.02;

        if (ticker.last >= targetPrice) {
          await waitForApiLimit();
          const order = await this.exchange.fetchOrder(state.position, symbol);
          const ethAmount = order.amount;
          await waitForApiLimit();
          const sellOrder = await this.exchange.createOrder(symbol, "market", "sell", ethAmount);
          console.log(`Sold ${ethAmount} ETH at ${ticker.last} USDT for a profit`);
          state.position = null;
          state.entryPrice = null;
        }
      }
    } catch (e) {
      console.error(`Error in runIteration: ${e.message}`);
      if (e.message.includes("Temporary lockout")) {
        console.error("Temporary lockout detected. Retrying in 60 seconds...");
        await new Promise((resolve) => setTimeout(resolve, 60000)); // Wait for 60 seconds before retrying
        this.runIteration(state, context, args); // Retry the function
      } else {
        console.error(`Unhandled exception: ${e.message}`);
      }
    }
  }
}

File: bot-6.py

# Buy ETH/USDT using 100% of starting capital. Sell at 2% profit

def initialize(self, state, context, args):
  state['position'] = None
  state['entry_price'] = None
  state['api_counter'] = 15  # Initialize with maximum allowed API counter
  state['last_api_call'] = time.time()  # Track the last API call time

def run_iteration(self, state, context, args):
  def wait_for_api_limit():
    elapsed_time = time.time() - state['last_api_call']
    state['api_counter'] += elapsed_time * 0.33  # Increment counter based on elapsed time
    state['api_counter'] = min(state['api_counter'], 15)  # Cap the counter at 15
    while state['api_counter'] < 1:  # Ensure at least one API call is available
        time.sleep(1)
        elapsed_time = time.time() - state['last_api_call']
        state['api_counter'] += elapsed_time * 0.33
        state['api_counter'] = min(state['api_counter'], 15)
    state['api_counter'] -= 1  # Decrement the counter for the current API call
    state['last_api_call'] = time.time()

  try:
      symbol = args.params['pair']
      capital_base = float(args.params['capitalBase'])
      wait_for_api_limit()
      balances = self.exchange.fetch_balance()
      usdt_balance = balances['USDT']['free']

      if state['position'] is None and usdt_balance >= capital_base:
          # Buy ETH/USDT using 100% of starting capital
          wait_for_api_limit()
          ticker = self.exchange.fetch_ticker(symbol)
          eth_amount = capital_base / ticker['last']
          wait_for_api_limit()
          order = self.exchange.create_order(symbol, 'market', 'buy', eth_amount)
          state['position'] = order['id']
          state['entry_price'] = ticker['last']
          log.info(f"Bought {eth_amount} ETH at {state['entry_price']} USDT")

      elif state['position'] is not None:
          # Check if we can sell at 2% profit
          wait_for_api_limit()
          ticker = self.exchange.fetch_ticker(symbol)
          target_price = state['entry_price'] * 1.02

          if ticker['last'] >= target_price:
              wait_for_api_limit()
              order = self.exchange.fetch_order(state['position'], symbol)
              eth_amount = order['amount']
              wait_for_api_limit()
              sell_order = self.exchange.create_order(symbol, 'market', 'sell', eth_amount)
              log.info(f"Sold {eth_amount} ETH at {ticker['last']} USDT for a profit")
              state['position'] = None
              state['entry_price'] = None

  except Exception as e:
        log.error(f"Error in run_iteration: {str(e)}")
        if "Temporary lockout" in str(e):
            log.error("Temporary lockout detected. Retrying in 60 seconds...")
            time.sleep(60)  # Wait for 60 seconds before retrying
            self.run_iteration(state, context, args)  # Retry the function
        else:
            log.error(f"Unhandled exception: {str(e)}")